src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorSourceUpdater.cs

```diff
@@ -48,59 +56,332 @@ public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, Imm
+            return new FileUpdaterResult(true, textReplacements.Select(r => r.OriginalText.FilePath).Distinct());
         }
 
-        public Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        private async Task<IList<TextReplacement>> GetReplacements(Project originalProject, IEnumerable<Document> updatedDocuments, CancellationToken token)
         {
-            throw new NotImplementedException();
+            List<TextReplacement> replacements = new List<TextReplacement>();
+            foreach (var updatedDocument in updatedDocuments)
+            {
+                var originalDocument = originalProject.GetDocument(updatedDocument.Id);
+                if (originalDocument is null)
+                {
+                    _logger.LogWarning("Cannot find expected generated file {FilePath}", updatedDocument.FilePath);
+                    continue;
+                }
+
+                // We want to correlate the mapped code blocks in the original document with corresponding ones in the updated document.
+                // Unfortunately, this is non-trivial because some code blocks may have been removed and, in other cases, multiple code blocks
+                // can have the same source location.
+                var originalSubTextGroups = (await MappedSubText.GetMappedSubTextsAsync(originalDocument, token).ConfigureAwait(false)).ToLookup(m => m.SourceLocation);
+                var updatedSubTextGroups = (await MappedSubText.GetMappedSubTextsAsync(updatedDocument, token).ConfigureAwait(false)).ToLookup(m => m.SourceLocation);
+
+                foreach (var originalGroup in originalSubTextGroups)
+                {
+                    var updatedGroup = updatedSubTextGroups[originalGroup.Key] ?? Enumerable.Empty<MappedSubText>();
+
+                    // If both groups have the same number of elements, then they pair in order
+                    if (originalGroup.Count() == updatedGroup.Count())
+                    {
+                        AddReplacementCandidates(replacements, originalGroup.Zip(updatedGroup, (original, updated) => new TextReplacement(original, updated.Text)));
+                    }
+
+                    // If the updated group is empty, then the original elements all pair with empty source text
+                    else if (!updatedGroup.Any())
+                    {
+                        AddReplacementCandidates(replacements, originalGroup.Select(m => new TextReplacement(m, SourceText.From(string.Empty))));
+                    }
+
+                    // This is the tricky one. If there are less updated code blocks than original code blocks, it will be necesary to guess which original code blocks
+                    // pair with the remaining updated code blocks based on text similarity.
+                    else
```

Comment:
how does this scenario come up? is this intended to handle having more than one code fixer per diagnostic?

If we only have one code fixer per diagnostic, and we're running one diagnostic at a time, then I would not expect overlapping code changes.

Reply:
The scenario is that some document has X pieces of C# that are mapped back to cshtml. After applying code fixers (could be one, could be more), we end up with X-1 pieces of C# (because one was removed entirely). The challenge is that we don't know which one was removed so it's tricky to map the updated C# to the original. I actually improved this logic in a newer commit, but it's still solving that problem.